### PR TITLE
Refactor: multi-phase module init + heap-allocated type

### PR DESCRIPTION
**Pull request by tseaver**
Similar to https://github.com/zopefoundation/zope.interface/pull/301, but **much** simpler:  there are no formerly static values to be moved to module state, and therefore none of the issues which make that PR not work for Python < 3.11.

### CODE REVIEW

File: src/zope/hookable/_zope_hookable.c [hunk 1]
```diff
@@ -12,153 +12,119 @@
...
-  if (self->implementation != NULL && visit(self->implementation, arg) < 0)
-    return -1;
-  if (self->old != NULL
-      && self->old != self->implementation
-      && visit(self->old, arg) < 0
-      )
-    return -1;
-
-  return 0;
+    Py_VISIT(Py_TYPE(self));
```
Comment (davisagli) [inline src/zope/hookable/_zope_hookable.c:57]: @tseaver https://docs.python.org/3.13/howto/isolating-extensions.html#tp-traverse-in-python-3-8-and-lower says this should not happen in Python 3.8 and lower.

**Review by davisagli — COMMENTED** [review]
For other reviewers: https://docs.python.org/3.13/howto/isolating-extensions.html gives some helpful context on the changes that are proposed here.

**Comment by davisagli** [comment]
Quick take on the failing builds:

1. 3.7 / macos: `Error: The version '3.7' with architecture 'arm64' was not found for macOS 14.4.1.` -> Probably needs to be set to an earlier version of the GHA macos image

2. 3.8 / macos: `error: call to undeclared function 'PyType_FromModuleAndSpec'` -> This function was added in 3.9. Huh, how come the other 3.8 builds were green?

3. 3.9 / manylinux aarch64: `ERROR: zope.hookable-6.1.dev0-0.editable-cp313-cp313t-linux_aarch64.whl is not a supported wheel on this platform.` -> not obvious to me where that comes from in quick skimming

4. various other builds in "waiting for status to be reported" state -> a maintainer probably needs to update which checks are required in the branch protection settings for this repository

File: src/zope/hookable/_zope_hookable.c (hunk 1 above)
Comment (tseaver) [reply src/zope/hookable/_zope_hookable.c:58]: @davisagli Thanks for the catch!  c3ef500

**Comment by tseaver** [comment]
@davisagli 
>3.8 / macos: error: call to undeclared function 'PyType_FromModuleAndSpec' -> This function was added in 3.9. Huh, how come the other 3.8 builds were green?

This PR needs an update like the one I figured out for `zope.i18nmessageid`, and applied to `zope.interface`:  do a `PY_VERSION_HEX` check for Python < 3.11 and do static type initalization.

I think the current tests pass because the package falls back silently to using the pure-Python implementation?

**Comment by tseaver** [comment]
@davisagli 
> 3.7 / macos: Error: The version '3.7' with architecture 'arm64' was not found for macOS 14.4.1. -> Probably needs to be set to an earlier version of the GHA macos image

Or even better, drop CI for 3.7 at all. ;)

**Comment by davisagli** [comment]
> Or even better, drop CI for 3.7 at all. ;)

I think it'd be reasonable to bump the major version and drop support for 3.7.

**Comment by tseaver** [comment]
@davisagli 
> I think it'd be reasonable to bump the major version and drop support for 3.7.

Looks like the [`meta` tooling](https://github.com/zopefoundation/meta/pull/217) is about ready for this step.

**Review by davisagli — APPROVED** [review]

**Pull request merged**